src/sagemaker_xgboost_container/checkpointing.py

```diff
@@ -0,0 +1,330 @@
+import logging
+import os
+import tempfile
+import threading
+import queue
+import re
+import xgboost as xgb
+from xgboost import rabit
+from xgboost.core import Booster, XGBoostError
+from xgboost.callback import _fmt_metric as format_metric
+
+
+TEMP_FILE_SUFFIX = ".sagemaker-ignore"
+FILE_LOCK_SUFFIX = ".sagemaker-uploading"
+FILE_SAFE_SUFFIX = ".sagemaker-uploaded"
+
+CHECKPOINT_FILENAME = "xgboost-checkpoint"
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+def train(train_args, checkpoint_dir):
+    """Convenience function for script mode.
+
+    Instead of running xgb.train(params, dtrain, ...), users can enable
+    checkpointing in script mode by creating a dictionary of xgb.train
+    arguments:
+
+    train_args = dict(
+        params=params,
+        dtrain=dtrain,
+        num_boost_round=num_round,
+        evals=[(dtrain, 'train'), (dtest, 'test')]
+    )
+
+    and calling:
+
+    bst = checkpointing.train(train_args)
+    """
+    train_args = train_args.copy()
+
+    xgb_model, start_iteration = load_checkpoint(checkpoint_dir)
+
+    if xgb_model is not None:
+        logging.info("Checkpoint loaded from %s", xgb_model)
+        logging.info("Resuming from iteration %s", start_iteration)
+
+    callbacks = train_args.get("callbacks", [])
+    callbacks.append(print_checkpointed_evaluation(start_iteration=start_iteration))
+    callbacks.append(save_checkpoint(checkpoint_dir, start_iteration=start_iteration))
+
+    train_args["verbose_eval"] = False  # suppress xgboost's print_evaluation()
+    train_args["xgb_model"] = xgb_model
+    train_args["callbacks"] = callbacks
+    # xgboost's default value for num_boost_round is 10.
+    num_boost_round = train_args.get("num_boost_round", 10) - start_iteration
+    # if last checkpoint is greater than num_boost_round, we shouldn't train.
+    train_args["num_boost_round"] = max(0, num_boost_round)
```

Comment:
We still need to call `xgb.train()` because `xgb_model` is a filename (str) while we have to return a booster object. But that line seems unnecessary. Removed it after testing `xgb.train()` with negative numbers and reading the source code upstream. Passing zero or negative number of rounds results in no training.